x/nft/base/handler.go

```diff
@@ -0,0 +1,161 @@
+package base
+
+import (
+	"fmt"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/x"
+	"github.com/iov-one/weave/x/nft"
+	"github.com/iov-one/weave/x/nft/blockchain"
+	"github.com/iov-one/weave/x/nft/ticker"
+	"github.com/iov-one/weave/x/nft/username"
+)
+
+const (
+	updateApprovalCost = 0
+)
+
+// RegisterRoutes will instantiate and register all handlers in this package
+func RegisterRoutes(r weave.Registry, auth x.Authenticator, issuer weave.Address) {
+	//TODO: might want to move that to init, this looks like an overhead we can avoid
+	bucketMap := map[string]Bucket{}
+	bucketMap[nft.Type_Username.String()] = Bucket{username.NewBucket().Bucket}
+	bucketMap[nft.Type_Ticker.String()] = Bucket{ticker.NewBucket().Bucket}
+	bucketMap[nft.Type_Blockchain.String()] = Bucket{blockchain.NewBucket().Bucket}
+
+	for _, v := range nft.Type_name {
+		if _, ok := bucketMap[v]; !ok {
+			panic(fmt.Sprintf("Bucket not registered in base nft handler: %s", v))
+		}
+	}
+
+	handler := NewApprovalOpsHandler(auth, issuer, bucketMap)
+	r.Handle(nft.PathAddApprovalMsg, handler)
+	r.Handle(nft.PathRemoveApprovalMsg, handler)
+}
+
+type Bucket struct {
+	orm.Bucket
+}
+
+type tokenHandler struct {
+	auth    x.Authenticator
+	issuer  weave.Address
+	buckets map[string]Bucket
+}
+
+// asBase will safely type-cast any value from Bucket
+func asBase(obj orm.Object) (nft.BaseNFT, error) {
+	if obj == nil || obj.Value() == nil {
+		return nil, nil
+	}
+	x, ok := obj.Value().(nft.BaseNFT)
+	if !ok {
+		return nil, nft.ErrUnsupportedTokenType()
+	}
+	return x, nil
+}
+
+func loadToken(typ nft.Type, h tokenHandler, store weave.KVStore, id []byte) (orm.Object, nft.BaseNFT, error) {
+	o, err := h.buckets[typ.String()].Get(store, id)
+	switch {
+	case err != nil:
+		return nil, nil, err
+	case o == nil:
+		return nil, nil, nft.ErrUnknownID()
+	}
+	t, e := asBase(o)
+	return o, t, e
+}
+
+func findActor(h tokenHandler, ctx weave.Context, t nft.BaseNFT) weave.Address {
+	if h.auth.HasAddress(ctx, t.OwnerAddress()) {
+		return t.OwnerAddress()
+	} else {
+		signers := x.GetAddresses(ctx, h.auth)
+		//TODO: revise, introduce updateApprovalsAction?
+		for _, signer := range signers {
+			if !t.Approvals().
+				List().
+				ForAction(nft.Action_ActionUpdateDetails.String()).
```

Comment:
Maybe `nft.Action_ActionUpdateDetails.String()` should be replaced by an argument?
so `findActorForUpdate()`, `findActorForBurn()`, etc could be wrappers like this and here add a fourth argument with and nft.Action value.

Reply:
You are right. We need to refactor that to use in all the occurrences. I haven't done it, because when I started this PR I built it per impl and in username impl we shared a common findActor.

I agree that this has to be moved to base level.